rotkehlchen/exchanges/bitcoinde.py

```diff
@@ -0,0 +1,249 @@
+from rotkehlchen.user_messages import MessagesAggregator
+from rotkehlchen.utils.misc import iso8601ts_to_timestamp
+from rotkehlchen.utils.serialization import rlk_jsonloads
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+def bitcoinde_asset(asset: str) -> Asset:
+    return Asset(asset.upper())
+
+
+def bitcoinde_pair_to_world(pair: str) -> Tuple[Asset, Asset]:
+    tx_asset = bitcoinde_asset(pair[:3])
+    native_asset = bitcoinde_asset(pair[3:])
+    return tx_asset, native_asset
+
+
+def trade_from_bitcoinde(raw_trade: Dict) -> Trade:
+
+    try:
+        timestamp = deserialize_timestamp_from_date(
+            raw_trade['successfully_finished_at'],
+            'iso8601',
+            'bitcoinde',
+        )
+    except KeyError:
+        # For very old trades (2013) bitcoin.de does not return 'successfully_finished_at'
+        timestamp = deserialize_timestamp_from_date(
+            raw_trade['trade_marked_as_paid_at'],
+            'iso8601',
+            'bitcoinde',
+        )
+
+    trade_type = deserialize_trade_type(raw_trade['type'])
+    tx_amount = AssetAmount(FVal(raw_trade['amount_currency_to_trade']))
+    native_amount = FVal(raw_trade['volume_currency_to_pay'])
+    tx_asset, native_asset = bitcoinde_pair_to_world(raw_trade['trading_pair'])
+    pair = TradePair(f'{tx_asset.identifier}_{native_asset.identifier}')
+    amount = tx_amount
+    rate = Price(native_amount / tx_amount)
+    fee_amount = deserialize_fee(raw_trade['fee_currency_to_pay'])
+    fee_asset = Asset('EUR')
+
+    return Trade(
+        timestamp=timestamp,
+        location=Location.BITCOINDE,
+        pair=pair,
+        trade_type=trade_type,
+        amount=amount,
+        rate=rate,
+        fee=fee_amount,
+        fee_currency=fee_asset,
+        link=str(raw_trade['trade_id']),
+    )
+
+
+class Bitcoinde(ExchangeInterface):
+    def __init__(
+            self,
+            api_key: ApiKey,
+            secret: ApiSecret,
+            database: 'DBHandler',
+            msg_aggregator: MessagesAggregator,
+    ):
+        super().__init__('bitcoinde', api_key, secret, database)
+        self.uri = 'https://api.bitcoin.de'
+        self.session.headers.update({'x-api-key': api_key})
+        self.msg_aggregator = msg_aggregator
+
+    def _generate_signature(self, request_type: str, url: str, nonce: str) -> str:
+        md5_empty = 'd41d8cd98f00b204e9800998ecf8427e'  # this is md5('')
+        signed_data = '#'.join([request_type, url, self.api_key, nonce, md5_empty]).encode()
+        signature = hmac.new(
+            self.secret,
+            signed_data,
+            hashlib.sha256,
+        ).hexdigest()
+        self.session.headers.update({
+            'x-api-signature': signature,
+        })
+        return signature
+
+    def _api_query(
+            self,
+            verb: str,
```

Comment:
you can actually even enforce this with types and mypy. Example: https://github.com/rotki/rotki/blob/4500381d5c527d255256f1a32d8261722ce028c2/rotkehlchen/exchanges/bittrex.py#L206